### PR TITLE
[gui] SceneV2 for a more efficient/cleaner implementation of Scene

### PR DESCRIPTION
**Pull request by AntonioFerreras**
Issue: #

### Brief Summary

Add a new scene object to the window. Can be retrieved using window.get_scene(). Replaces using ti.ui.Scene() to directly instantiate scene object. 

The new scenev2 implementation changes the way data of renderables and scene info are uploaded to the GPU for efficiency. This PR incudes the code changes and updates the ggui examples, but does not update tests or documentation.
### <samp>🤖 Generated by Copilot at a3e01bb</samp>

*  Refactor the scene API to support a new `SceneV2` class that provides more rendering features and flexibility for 3D objects. ( [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-de0a9a1fb6c6210ac5653025b79a982b6721b1c66e17d2dca1db90fa82613bddR455-R458), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-de0a9a1fb6c6210ac5653025b79a982b6721b1c66e17d2dca1db90fa82613bddR701-R709), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-15288b9648dcdf9969af4bc516af6a08c3529112b74ba4524c627b5ba411626fR1-R34), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-5d0ee39a6fca82325fe07b5491d7ac6b7367ec675730dd76dfc4909e6969321aR9))
*  Deprecate the direct instantiation of `ti.ui.Scene` and suggest using the `window.get_scene()` method instead, which returns a `SceneV2` instance. ([link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-74259ee62a5b2862ca7cd1b9e8590901d74f47cc51d576671d1fb610b1eae8b0L569-R569), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-7bfd13e221e34c8de05c26a2cec3a75278097b3ee63e59f71c83bb648ce692d9L128-R128), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-c180380281cbad6c17bf08be1986c785e0ac20b22f96f35eeda90091498a0642L245-R245), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-fa83b5cb7437573b71ca34dd6925d872a8c6aebc383919b552045a5bfa95a3bcL377-R377), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-3db65eaf8189bc001eef73837950994805c91751c2ab9d0d2423a4cc61ebfa21R88-R91), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-4045871273e551070e37df548b65f80426025c5c45e14dad2dc0f862c3e33f95R122-R125), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-de0a9a1fb6c6210ac5653025b79a982b6721b1c66e17d2dca1db90fa82613bddR621-R625))
*  Integrate the new scene API with the existing canvas API, allowing the `canvas.scene` method to accept either a `ti.ui.Scene` or a `ti.ui.SceneV2` instance. ([link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-4f7ced603627bc0dac36eae499062271ca7f5b0f3724c688f758950802d7136dR4), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-4f7ced603627bc0dac36eae499062271ca7f5b0f3724c688f758950802d7136dR244-R246), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-4045871273e551070e37df548b65f80426025c5c45e14dad2dc0f862c3e33f95R16), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-de0a9a1fb6c6210ac5653025b79a982b6721b1c66e17d2dca1db90fa82613bddL525-R679), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-8fdac82731b23fd9ef6c81a34d21eaae92d89afd438815cced6ea6bbb052be18L38-R42), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-9836f057d9e87aafad628d30f61876ae58afcef5e5dac26ea5bfde89cacc0afcR23))
*  Implement the `update_scene_data` method for the renderable classes that need scene-related data, such as the light sources and the camera. ([link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-de0a9a1fb6c6210ac5653025b79a982b6721b1c66e17d2dca1db90fa82613bddR292-R434), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-872f7266d3398a5b2b83e699974f22d519df20eb70a94b6301db365dd83c3f50R192-R196), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-8d490fe29c2c71ac75dd2156a0505c2af37337ffff689bff9aabab91efab7ef1L56-R58), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-e55ebf251f3a37554b63662ec6d681fa76faa04fd236e17794540f19246f90bfL21-R21), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-791de621f7e9beba1ecf8b1df79d3e823f384b498881ed3e3982b6d954788896L33-R33), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-cfad5db6c53643971b9c8a40ad082ea715de90bb5327e0f186fe94e5b45d2572R22), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-d54f4426461756f88ef5bc0a4f14f0b6c0704aeadc5c37d804fee447bab32e50L34-R34), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-6b04c9d6caf02eb6b4dedf3889a3216a25b1ba768a28bddfac14c638e0faedf4R22), [link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-6b04c9d6caf02eb6b4dedf3889a3216a25b1ba768a28bddfac14c638e0faedf4L55-R57))
*  Add an import statement for the `warnings` module in `python/taichi/ui/scene.py`. ([link](https://github.com/taichi-dev/taichi/pull/8205/files?diff=unified&w=0#diff-3db65eaf8189bc001eef73837950994805c91751c2ab9d0d2423a4cc61ebfa21R1))

### CODE REVIEW

File: taichi/ui/backends/vulkan/canvas.h [hunk 1]
```diff
@@ -20,6 +20,7 @@
 #include "taichi/ui/backends/vulkan/swap_chain.h"
 #include "taichi/ui/backends/vulkan/renderable.h"
 #include "taichi/ui/common/canvas_base.h"
+#include "taichi/ui/backends/vulkan/sceneV2.h"
```
Comment (ailzhang) [inline taichi/ui/backends/vulkan/canvas.h:23]: nit: this can be possibly moved to `canvas.cpp`?

File: python/taichi/ui/scene.py [hunk 2]
```diff
@@ -84,6 +434,10 @@ class Scene:
     def __init__(self):
         check_ggui_availability()
         self.scene = _ti_core.PyScene()
+        warnings.warn(
```
Comment (ailzhang) [inline python/taichi/ui/scene.py:437]: Please add a test in `test_deprecation.py" so that we can keep track of the removal of deprecated apis ;)

**Review by ailzhang — APPROVED** [review]
Looks great! Thanks!

File: taichi/ui/backends/vulkan/canvas.h (hunk 1 above)
Comment (AntonioFerreras) [reply taichi/ui/backends/vulkan/canvas.h]: changed :)

File: python/taichi/ui/scene.py (hunk 2 above)
Comment (AntonioFerreras) [reply python/taichi/ui/scene.py:437]: will add in next PR that adds tests for sceneV2

**Pull request merged**